### PR TITLE
redfish_storage_volume resource added

### PR DESCRIPTION
**Pull request by mikeletux**
##### SUMMARY

This pull request adds a basic resource to **manage volumes** using the redfish API.

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

resource_redfish_storage_volume

##### ADDITIONAL INFORMATION

This module allows users to create and destroy volumes on servers using the redfish API.
Currently it supports creation and deletion of volumes immediate (synchronous) and on reset (asynchronous, needs reboot).  
Synchronous is always preferred.
The behavior is the following:
When Immediate is selected, terraform waits till the volume is created. It also waits until the volume is destroyed.
When OnReset is selected, it creates the task and finishes. When destroying, if the job has not been executed yet, it kills the job. If the server has been rebooted and the changes applied, it kills the volume itself.

**ATTENTION - IMPORTANT LIMITATION**
Since this pull request has been developed with an iDRAC (Dell implementation of Redfish API), it still doesn't support HTTP DELETE on tasks or taskmonitor endpoints. This lead me to write a method (common.DeleteDellJob()) to remove jobs when needed. This uses the specific iDRAC calls, which won't be working in other vendors. As soon as this is implemented in iDRAC, I will use the standard redfish call for deleting tasks.

Also since disks and volumes are something no one would update (volumes would be deleted an recreated) methods update and read do nothing here. Please do not try to update the resource since it will lead to inconsistencies between the state file and the actual state of the resource.

**tf file**

```paste below
provider "redfish" {
  redfish_endpoint = "https://localhost:5000"
  user = "root"
  password = "calvin"
  ssl_insecure = true
}

resource "redfish_storage_volume" "volume" {
    storage_controller = "RAID.Integrated.1-1"
    volume_name = "MyVol"
    raid_level = "Mirrored"
    volume_disks = ["Physical Disk 0:1:0", "Physical Disk 0:1:1"]
    settings_apply_time = "Immediate"
//    settings_apply_time = "OnReset"
}
```
**Immediate execution**
```paste below
[mikeletux@miguel-virtualbox storage]$ terraform apply

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # redfish_storage_volume.volume will be created
  + resource "redfish_storage_volume" "volume" {
      + bios_config_job_uri = (known after apply)
      + id                  = (known after apply)
      + raid_level          = "Mirrored"
      + settings_apply_time = "Immediate"
      + storage_controller  = "RAID.Integrated.1-1"
      + volume_disks        = [
          + "Physical Disk 0:1:0",
          + "Physical Disk 0:1:1",
        ]
      + volume_name         = "MyVol"
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

redfish_storage_volume.volume: Creating...
redfish_storage_volume.volume: Still creating... [10s elapsed]
redfish_storage_volume.volume: Still creating... [20s elapsed]
redfish_storage_volume.volume: Still creating... [30s elapsed]
redfish_storage_volume.volume: Still creating... [40s elapsed]
redfish_storage_volume.volume: Still creating... [50s elapsed]
redfish_storage_volume.volume: Still creating... [1m0s elapsed]
redfish_storage_volume.volume: Still creating... [1m10s elapsed]
redfish_storage_volume.volume: Still creating... [1m20s elapsed]
redfish_storage_volume.volume: Still creating... [1m30s elapsed]
redfish_storage_volume.volume: Still creating... [1m40s elapsed]
redfish_storage_volume.volume: Still creating... [1m50s elapsed]
redfish_storage_volume.volume: Creation complete after 1m59s [id=/redfish/v1/Systems/System.Embedded.1/Storage/RAID.Integrated.1-1/Volumes/Disk.Virtual.0:RAID.Integrated.1-1]
```
**OnReset execution**
```
[mikeletux@miguel-virtualbox storage]$ terraform apply

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # redfish_storage_volume.volume will be created
  + resource "redfish_storage_volume" "volume" {
      + bios_config_job_uri = (known after apply)
      + id                  = (known after apply)
      + raid_level          = "Mirrored"
      + settings_apply_time = "OnReset"
      + storage_controller  = "RAID.Integrated.1-1"
      + volume_disks        = [
          + "Physical Disk 0:1:0",
          + "Physical Disk 0:1:1",
        ]
      + volume_name         = "MyVol"
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

redfish_storage_volume.volume: Creating...
redfish_storage_volume.volume: Still creating... [10s elapsed]
redfish_storage_volume.volume: Creation complete after 12s [id=/redfish/v1/TaskService/Tasks/JID_013875086993]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
```
***Check the differences in ID and execution time between Immediate and OnReset***

#### Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
FEATURES:
* **New Resource** `redfish_storage_volume`
```

#### Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccXXX'

...
```

#### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

### CODE REVIEW

File: examples/storage/volume.tf [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  user = "root"
+  password = "calvin"
+  ssl_insecure = true
+}
+
+resource "redfish_storage_volume" "volume" {
+    storage_controller = "RAID.Integrated.1-1"
+    volume_name = "MyVol"
+    raid_level = "Mirrored"
+    volume_disks = ["Physical Disk 0:1:0", "Physical Disk 0:1:1"]
```
Comment (anupamaloke) [inline examples/storage/volume.tf:12]: These don't look like valid drive IDs, at least for iDRAC. Usually, these are of form ```Disk.Direct.0-0:AHCI.Embedded.1-1``` or ```Disk.Bay.0:Enclosure.Internal.0-1:RAID.Slot.1-1```.

File: go.mod [hunk 2]
```diff
@@ -5,5 +5,5 @@ go 1.14
 require (
 	github.com/hashicorp/terraform-plugin-sdk v1.14.0
 	github.com/hashicorp/terraform-plugin-sdk/v2 v2.0.1
-	github.com/stmcginnis/gofish v0.5.1-0.20200820193452-a93a78ebb0cf
+	github.com/stmcginnis/gofish v0.5.0
```
[the remark this answers is not in the document]
Comment (mikeletux) [reply go.mod:8]: Well, actually version v0.5.0 was the latest one. 
There has been new commits from them but Sean has not created a tag to them, so the way Go handles that case is putting a +v0.0.1 to the current version DOT the current time DASH the 12 firsts octets of the commit.
We have to discuss how to handle this.

File: redfish/resource_redfish_storage_volume.go [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+			},
+			settingsApplyTimeStr: &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			biosConfigJobURIStr: &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			/*ValidateFunc: xxxx,
```
[the remark this answers is not in the document]
Comment (mikeletux) [reply redfish/resource_redfish_storage_volume.go:63]: Yes, I agree, I'll change that

File: redfish/resource_redfish_storage_volume.go [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+				Description: "This value must be the disk controller the user want to manage. I.e: RAID.Integrated.1-1",
+			},
+			volumeNameStr: &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "This value is the desired name for the volume to be given",
+			},
+			raidLevelStr: &schema.Schema{ //Call it volumeType
+				Type:        schema.TypeString,
+				Required:    true,
```
[the remark this answers is not in the document]
Comment (mikeletux) [reply redfish/resource_redfish_storage_volume.go:44]: I agree, but I think with terraform we assume that the resource is not created in advance, so first we need to create it (that's why I set the raidLevel to required, so we know what to create).
To destroy, once created, is just a matter of "terraform destroy". Am I right?

File: redfish/resource_redfish_storage_volume.go [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+}
+
+func resourceStorageVolumeCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	var diags diag.Diagnostics
+	conn := m.(*gofish.APIClient)
+	//Get user config
+	storageController := d.Get(storageControllerStr).(string)
+	raidLevel := d.Get(raidLevelStr).(string)
+	volumeName := d.Get(volumeNameStr).(string)
+	applyTime := d.Get(settingsApplyTimeStr).(string)
```
[the remark this answers is not in the document]
Comment (mikeletux) [reply redfish/resource_redfish_storage_volume.go:80]: Yes, totally agree, I'll change that

**Comment by mikeletux** [comment]
Hi @anupamaloke ,
I've done the changes discuss in our weekly call.

Please check and if they look good, approve the PR 😄 
/Miguel

**Comment by mikeletux** [comment]
@anupamaloke feel free to merge.
I've added some tests for the storage resource.
I've also created a function to create step by step the mocked client. We can discuss that in our next weekly call.

Thanks!

**Pull request merged**